File: src/core/typedb.cpp
#include "typedb.h"

#include <algorithm>
#include <functional>

namespace oxi {
namespace {

bool starts_with(const std::string& s, const std::string& prefix) {
    return s.size() >= prefix.size() && s.compare(0, prefix.size(), prefix) == 0;
}

RustTypePtr make(RKind kind, long long size = 0) {
    auto t = std::make_shared<RustType>();
    t->kind = kind;
    t->size = size;
    return t;
}

RustTypePtr make_int(long long size, int bits, bool is_signed, const std::string& name) {
    auto t = make(RKind::Int, size);
    t->bits = bits;
    t->is_signed = is_signed;
    t->prim_name = name;
    return t;
}

long long variant_payload_size(const RustVariant& v) {
    long long s = 0;
    for (const auto& f : v.fields) {
        if (f.ty) s += f.ty->size;
    }
    return s;
}

}  // namespace

bool RustType::is_option() const {
    if (kind != RKind::Enum || !starts_with(name, "core::option::Option")) return false;
    std::set<std::string> names;
    for (const auto& v : variants) names.insert(v.name);
    return names == std::set<std::string>{"Some", "None"};
}

bool RustType::is_result() const {
    if (kind != RKind::Enum || !starts_with(name, "core::result::Result")) return false;
    std::set<std::string> names;
    for (const auto& v : variants) names.insert(v.name);
    return names == std::set<std::string>{"Ok", "Err"};
}

bool RustType::is_niche_encoded() const {
    if (kind != RKind::Enum) return false;
    long long payload = 0;
    for (const auto& v : variants) payload = std::max(payload, variant_payload_size(v));
    return payload >= size && size > 0;
}

TypeDB TypeDB::from_file(const std::string& path, int arch_bits) {
    return from_json(json_parse_file(path), arch_bits);
}

TypeDB TypeDB::from_json(const JsonValue& data, int arch_bits) {
    TypeDB db;
    db.arch_bits = arch_bits;
    db.arch_bytes = arch_bits / 8;
    db.load(data);
    return db;
}

void TypeDB::load(const JsonValue& data) {
    const JsonValue& structs = data["structs"];
    for (const auto& s : structs.items()) {
        if (s.is_object() && s.has("name")) {
            struct_db_[s["name"].as_string()] = &s;
        }
    }
    for (const auto& s : structs.items()) {
        parse_type(s);
    }
    const JsonValue& functions = data["functions"];
    for (const auto& f : functions.items()) {
        RustFnProto proto;
        if (parse_prototype(f["prototype"], proto)) {
            prototypes[f["name"].as_string()].push_back(std::move(proto));
        }
    }
}

RustTypePtr TypeDB::parse_type(const JsonValue& data) {
    if (!data.is_object()) return nullptr;
    std::string kind = data["kind"].as_string("");
    if (kind == "Pointer") return parse_pointer(data);
    if (kind == "Primitive") return parse_primitive(data);
    if (kind == "Struct") return parse_struct(data);
    if (kind == "Enumeration") return parse_enum(data);
    if (kind == "Array") return parse_array(data);
    if (kind == "None") return nullptr;
    return nullptr;
}

RustTypePtr TypeDB::parse_pointer(const JsonValue& data) {
    RustTypePtr pts;
    if (data.has("pts_to") && !data["pts_to"].is_null()) {
        pts = parse_type(data["pts_to"]);
    }
    if (!pts) pts = make(RKind::Bottom);
    auto t = make(RKind::Pointer, arch_bytes);
    t->pts_to = pts;
    return t;
}

RustTypePtr TypeDB::parse_primitive(const JsonValue& data) {
    std::string name = data["name"].as_string("");
    if (name == "i8") return make_int(1, 8, true, "i8");
    if (name == "i16") return make_int(2, 16, true, "i16");
    if (name == "i32") return make_int(4, 32, true, "i32");
    if (name == "i64") return make_int(8, 64, true, "i64");
    if (name == "i128") return make_int(16, 128, true, "i128");
    if (name == "u8") return make_int(1, 8, false, "u8");
    if (name == "u16") return make_int(2, 16, false, "u16");
    if (name == "u32") return make_int(4, 32, false, "u32");
    if (name == "u64") return make_int(8, 64, false, "u64");
    if (name == "u128") return make_int(16, 128, false, "u128");
    if (name == "f32") { auto t = make(RKind::Float, 4); t->bits = 32; return t; }
    if (name == "f64") { auto t = make(RKind::Float, 8); t->bits = 64; return t; }
    if (name == "bool") return make(RKind::Bool, 1);
    if (name == "char") return make(RKind::Char, 4);
    long long size = data["size"].as_int(0);
    if (size > 0) return make_int(size, static_cast<int>(size * 8), false, name);
    return nullptr;
}

RustTypePtr TypeDB::parse_array(const JsonValue& data) {
    RustTypePtr ele = parse_type(data["ele_type"]);
    if (!ele) return nullptr;
    long long length = data["length"].as_int(0);
    auto t = make(RKind::Array, ele->size * length);
    t->ele = ele;
    t->length = length;
    return t;
}

RustTypePtr TypeDB::parse_struct(const JsonValue& data) {
    std::string name = data["name"].as_string("");
    auto existing = types.find(name);
    if (existing != types.end()) return existing->second;
    if (pending_.count(name)) return make(RKind::Bottom);

    const JsonValue& fields_data = data["fields"];
    if (fields_data.is_null()) {
        if (resolving_.count(name)) return nullptr;
        auto it = struct_db_.find(name);
        if (it == struct_db_.end() || it->second == &data) return nullptr;
        resolving_.insert(name);
        RustTypePtr r = parse_struct(*it->second);
        resolving_.erase(name);
        return r;
    }

    pending_.insert(name);
    std::vector<RustField> fields;
    bool ok = true;
    for (const auto& member : fields_data.members()) {
        // member.second is [field_name, field_type]
        const JsonValue& pair = member.second;
        if (!pair.is_array() || pair.size() < 2) { ok = false; break; }
        RustTypePtr fty = parse_type(pair.items()[1]);
        if (!fty) { ok = false; break; }
        RustField fld;
        try {
            fld.offset = std::stoll(member.first);
        } catch (...) {
            fld.offset = 0;
        }
        fld.name = pair.items()[0].as_string("");
        fld.ty = fty;
        fields.push_back(std::move(fld));
    }
    pending_.erase(name);
    if (!ok) return nullptr;

    auto result = make(RKind::Struct, data["size"].as_int(0));
    result->name = name;
    result->fields = std::move(fields);
    RustTypePtr patched = apply_patches(result);
    types[name] = patched;
    return patched;
}

RustTypePtr TypeDB::parse_enum(const JsonValue& data) {
    std::string name = data["name"].as_string("");
    auto existing = types.find(name);
    if (existing != types.end()) return existing->second;
    if (pending_.count(name)) return make(RKind::Bottom);

    const JsonValue& variants_data = data["variants"];
    if (variants_data.is_null()) {
        if (resolving_.count(name)) return nullptr;
        auto it = struct_db_.find(name);
        if (it == struct_db_.end() || it->second == &data) return nullptr;
        resolving_.insert(name);
        RustTypePtr r = parse_enum(*it->second);
        resolving_.erase(name);
        return r;
    }

    pending_.insert(name);
    long long disc_size = data["discriminant_size"].as_int(0);
    std::vector<RustVariant> variants;
    bool ok = true;
    for (const auto& member : variants_data.members()) {
        const JsonValue& vd = member.second;  // [discriminant, [[fname, ftype], ...]]
        if (!vd.is_array() || vd.size() < 2) { ok = false; break; }
        const JsonValue& disc = vd.items()[0];
        const JsonValue& vfields = vd.items()[1];
        RustVariant var;
        var.name = member.first;
        var.has_disc = !disc.is_null();
        var.discriminant = disc.as_int(0);
        var.disc_size = var.has_disc ? disc_size : 0;
        for (const auto& fld : vfields.items()) {
            if (!fld.is_array() || fld.size() < 2) { ok = false; break; }
            RustTypePtr fty = parse_type(fld.items()[1]);
            if (!fty) { ok = false; break; }
            RustField f;
            f.offset = 0;  // payload-relative; recomputed on lowering
            f.name = fld.items()[0].as_string("");
            f.ty = fty;
            var.fields.push_back(std::move(f));
        }
        if (!ok) break;
        variants.push_back(std::move(var));
    }
    pending_.erase(name);
    if (!ok) return nullptr;

    auto result = make(RKind::Enum, data["size"].as_int(0));
    result->name = name;
    result->variants = std::move(variants);
    result->disc_size = disc_size;
    types[name] = result;
    return result;
}

RustTypePtr TypeDB::apply_patches(const RustTypePtr& ty) {
    RustTypePtr t = to_slice(ty);
    if (t->kind == RKind::Struct) t = unwrap_argument(t);
    return t;
}

RustTypePtr TypeDB::to_slice(const RustTypePtr& ty) {
    std::set<std::string> names;
    for (const auto& f : ty->fields) names.insert(f.name);
    if (names != std::set<std::string>{"data_ptr", "length"}) return ty;
    RustTypePtr data_ptr, length;
    for (const auto& f : ty->fields) {
        if (f.name == "data_ptr") data_ptr = f.ty;
        if (f.name == "length") length = f.ty;
    }
    if (data_ptr && data_ptr->kind == RKind::Pointer && length && length->kind == RKind::Int &&
        length->size == arch_bytes) {
        if (ty->name == "&str" && data_ptr->pts_to && data_ptr->pts_to->kind == RKind::Int &&
            data_ptr->pts_to->bits == 8) {
            return make(RKind::StrRef, ty->size);
        }
        auto slice = make(RKind::Slice, ty->size);
        slice->ele = data_ptr->pts_to;
        return slice;
    }
    return ty;
}

RustTypePtr TypeDB::unwrap_argument(const RustTypePtr& ty) {
    if (ty->name != "core::fmt::rt::Argument") return ty;
    std::set<std::string> names;
    for (const auto& f : ty->fields) names.insert(f.name);
    if (names != std::set<std::string>{"ty"}) return ty;
    RustTypePtr inner = ty->fields[0].ty;
    if (inner && inner->kind == RKind::Enum) {
        for (const auto& v : inner->variants) {
            if (v.name == "Placeholder") {
                auto s = make(RKind::Struct, ty->size);
                s->name = "core::fmt::rt::Argument";
                s->fields = v.fields;
                return s;
            }
        }
    }
    return ty;
}

bool TypeDB::parse_prototype(const JsonValue& data, RustFnProto& out) {
    if (!data.is_object()) return false;
    std::vector<RustTypePtr> args;
    for (const auto& a : data["args"].items()) {
        RustTypePtr t = parse_type(a);
        if (!t) return false;
        args.push_back(t);
    }
    RustTypePtr ret = parse_type(data["returnty"]);
    out = fit_abi(std::move(args), ret);
    return true;
}

RustFnProto TypeDB::fit_abi(std::vector<RustTypePtr> args, RustTypePtr ret) {
    long long big = static_cast<long long>(arch_bytes) * 2;
    auto is_aggregate = [](const RustTypePtr& t) {
        return t && (t->kind == RKind::Struct || t->kind == RKind::Enum);
    };
    RustFnProto proto;
    for (auto& a : args) {
        if (is_aggregate(a) && a->size > big) {
            auto p = make(RKind::Pointer, arch_bytes);
            p->pts_to = a;
            proto.args.push_back(p);
        } else {
            proto.args.push_back(a);
        }
    }
    if (is_aggregate(ret) && ret->size > big) {
        auto p = make(RKind::Pointer, arch_bytes);
        p->pts_to = ret;
        proto.args.insert(proto.args.begin(), p);
        proto.ret = nullptr;
        proto.is_retbuf = true;
    } else {
        proto.ret = ret;
        proto.is_retbuf = false;
    }
    return proto;
}

}  // namespace oxi

File: reference_py/oxidizer_core/typedb.py
"""Parser for Oxidizer's per-version Rust type database (``type_db/<ver>.json``).

This is an IDA-independent reimplementation of the parsing half of
``angr/rust/analyses/type_db_loader.py``.  It turns the JSON into a small typed
IR (the ``Rust*`` dataclasses below); the ``oxidizer_core.ida_types`` adapter
then lowers that IR onto Hex-Rays ``tinfo_t`` objects.

Differences from the angr version, all deliberate improvements for IDA output
(angr discarded these because its renderer could not use them):

* ``f32``/``f64`` map to real floats instead of being dropped (angr returned
  ``None``, which silently rejected every struct containing a float).
* ``char`` maps to a 4-byte value (Rust ``char`` is a Unicode scalar) instead of
  a single byte.

All sizes in the JSON — and therefore in this IR — are in **bytes**.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path


# ---------------------------------------------------------------------------
# IR
# ---------------------------------------------------------------------------


@dataclass
class RustType:
    size: int = 0

    def render(self) -> str:  # pragma: no cover - overridden
        return "?"


@dataclass
class RustInt(RustType):
    bits: int = 32
    signed: bool = False
    name: str = ""

    def render(self) -> str:
        if self.name:
            return self.name
        return f"{'i' if self.signed else 'u'}{self.bits}"


@dataclass
class RustFloat(RustType):
    bits: int = 64

    def render(self) -> str:
        return f"f{self.bits}"


@dataclass
class RustBool(RustType):
    def render(self) -> str:
        return "bool"


@dataclass
class RustChar(RustType):
    def render(self) -> str:
        return "char"


@dataclass
class RustPointer(RustType):
    pts_to: RustType | None = None

    def render(self) -> str:
        return f"*{self.pts_to.render() if self.pts_to else '()'}"


@dataclass
class RustArray(RustType):
    ele: RustType | None = None
    length: int = 0

    def render(self) -> str:
        return f"[{self.ele.render() if self.ele else '?'}; {self.length}]"


@dataclass
class RustSlice(RustType):
    ele: RustType | None = None

    def render(self) -> str:
        return f"&[{self.ele.render() if self.ele else '?'}]"


@dataclass
class RustStrRef(RustType):
    def render(self) -> str:
        return "&str"


@dataclass
class RustField:
    offset: int
    name: str
    ty: RustType


@dataclass
class RustStruct(RustType):
    name: str = ""
    fields: list[RustField] = field(default_factory=list)

    def render(self) -> str:
        return self.name or "struct"


@dataclass
class RustVariant:
    name: str
    discriminant: int | None
    discriminant_size: int
    fields: list[RustField] = field(default_factory=list)


@dataclass
class RustEnum(RustType):
    name: str = ""
    variants: list[RustVariant] = field(default_factory=list)
    discriminant_size: int = 0

    @property
    def is_option(self) -> bool:
        return self.name.startswith("core::option::Option") and {v.name for v in self.variants} == {"Some", "None"}

    @property
    def is_result(self) -> bool:
        return self.name.startswith("core::result::Result") and {v.name for v in self.variants} == {"Ok", "Err"}

    @property
    def is_niche_encoded(self) -> bool:
        """True when the enum has no room for a discriminant tag (the payload's
        invalid bit-patterns encode the other variant, e.g. ``Option<&T>``)."""
        payload = max((self._variant_size(v) for v in self.variants), default=0)
        return payload >= self.size > 0

    def _variant_size(self, v: RustVariant) -> int:
        return sum(f.ty.size for f in v.fields)

    def render(self) -> str:
        return self.name or "enum"


@dataclass
class RustBottom(RustType):
    """Placeholder for an in-progress (recursive) type."""

    def render(self) -> str:
        return "_"


@dataclass
class RustFnProto:
    args: list[RustType]
    ret: RustType | None
    is_retbuf: bool = False

    def render(self) -> str:
        a = ", ".join(t.render() for t in self.args)
        if self.is_retbuf:
            return f"fn({a})  /* arg0 = return buffer */"
        return f"fn({a}) -> {self.ret.render() if self.ret else '()'}"


# ---------------------------------------------------------------------------
# parser
# ---------------------------------------------------------------------------

_PRIMITIVES = {
    "i8": lambda: RustInt(1, 8, True, "i8"),
    "i16": lambda: RustInt(2, 16, True, "i16"),
    "i32": lambda: RustInt(4, 32, True, "i32"),
    "i64": lambda: RustInt(8, 64, True, "i64"),
    "i128": lambda: RustInt(16, 128, True, "i128"),
    "u8": lambda: RustInt(1, 8, False, "u8"),
    "u16": lambda: RustInt(2, 16, False, "u16"),
    "u32": lambda: RustInt(4, 32, False, "u32"),
    "u64": lambda: RustInt(8, 64, False, "u64"),
    "u128": lambda: RustInt(16, 128, False, "u128"),
    "f32": lambda: RustFloat(4, 32),
    "f64": lambda: RustFloat(8, 64),
    "bool": lambda: RustBool(1),
    "char": lambda: RustChar(4),
}


class TypeDB:
    """Loads and parses a single ``type_db/<version>.json`` file."""

    def __init__(self, arch_bits: int = 64):
        self.arch_bits = arch_bits
        self.arch_bytes = arch_bits // 8
        self.types: dict[str, RustType] = {}  # struct/enum name -> IR
        self.prototypes: dict[str, list[RustFnProto]] = {}
        self._struct_db: dict[str, dict] = {}
        self._pending: set[str] = set()
        self._resolving: set[str] = set()  # guards null-definition forwarding cycles

    # -- entry points ------------------------------------------------------
    @classmethod
    def from_file(cls, path: str | Path, arch_bits: int = 64) -> "TypeDB":
        return cls.from_json(json.loads(Path(path).read_text()), arch_bits)

    @classmethod
    def from_json(cls, data: dict, arch_bits: int = 64) -> "TypeDB":
        db = cls(arch_bits)
        db._load(data)
        return db

    def _load(self, data: dict) -> None:
        self._struct_db = {s["name"]: s for s in data.get("structs", [])}
        for sdata in self._struct_db.values():
            self._parse_type(sdata)
        for fdata in data.get("functions", []):
            proto = self._parse_prototype(fdata.get("prototype"))
            if proto is not None:
                self.prototypes.setdefault(fdata["name"], []).append(proto)

    # -- dispatch ----------------------------------------------------------
    def _parse_type(self, data) -> RustType | None:
        if not isinstance(data, dict):
            return None
        kind = data.get("kind")
        if kind == "Pointer":
            return self._parse_pointer(data)
        if kind == "Primitive":
            return self._parse_primitive(data)
        if kind == "Struct":
            return self._parse_struct(data)
        if kind == "Enumeration":
            return self._parse_enum(data)
        if kind == "Array":
            return self._parse_array(data)
        if kind == "None":
            return None
        return None

    def _parse_pointer(self, data) -> RustPointer:
        pts = self._parse_type(data["pts_to"]) if data.get("pts_to") else None
        return RustPointer(size=self.arch_bytes, pts_to=pts or RustBottom())

    def _parse_primitive(self, data) -> RustType | None:
        name = data.get("name", "")
        ctor = _PRIMITIVES.get(name)
        if ctor is not None:
            return ctor()
        size = data.get("size")
        if size:
            return RustInt(size=size, bits=size * 8, signed=False, name=name)
        return None

    def _parse_array(self, data) -> RustArray | None:
        ele = self._parse_type(data["ele_type"])
        if ele is None:
            return None
        length = data.get("length", 0)
        return RustArray(size=ele.size * length, ele=ele, length=length)

    def _parse_struct(self, data) -> RustType | None:
        name = data["name"]
        if name in self.types:
            return self.types[name]
        if name in self._pending:
            return RustBottom()
        fields_data = data.get("fields")
        if fields_data is None:
            if name in self._resolving:
                return None
            actual = self._struct_db.get(name)
            if not actual or actual is data:
                return None
            self._resolving.add(name)
            try:
                return self._parse_struct(actual)
            finally:
                self._resolving.discard(name)
        self._pending.add(name)
        fields: list[RustField] = []
        ok = True
        for off_key, (fname, fdata) in fields_data.items():
            fty = self._parse_type(fdata)
            if fty is None:
                ok = False
                break
            fields.append(RustField(int(off_key), fname, fty))
        self._pending.discard(name)
        if not ok:
            return None
        result: RustType = RustStruct(size=data.get("size", 0), name=name, fields=fields)
        result = self._apply_patches(result)
        self.types[name] = result
        return result

    def _parse_enum(self, data) -> RustType | None:
        name = data["name"]
        if name in self.types:
            return self.types[name]
        if name in self._pending:
            return RustBottom()
        variants_data = data.get("variants")
        if variants_data is None:
            if name in self._resolving:
                return None
            actual = self._struct_db.get(name)
            if not actual or actual is data:
                return None
            self._resolving.add(name)
            try:
                return self._parse_enum(actual)
            finally:
                self._resolving.discard(name)
        self._pending.add(name)
        disc_size = data.get("discriminant_size", 0)
        variants: list[RustVariant] = []
        ok = True
        for vname, (disc, vfields_data) in variants_data.items():
            vfields: list[RustField] = []
            for fname, fdata in vfields_data:
                fty = self._parse_type(fdata)
                if fty is None:
                    ok = False
                    break
                # variant field offsets are payload-relative; recompute on lowering
                vfields.append(RustField(0, fname, fty))
            if not ok:
                break
            variants.append(RustVariant(vname, disc, disc_size if disc is not None else 0, vfields))
        self._pending.discard(name)
        if not ok:
            return None
        result = RustEnum(size=data.get("size", 0), name=name, variants=variants, discriminant_size=disc_size)
        self.types[name] = result
        return result

    # -- patches (ported from type_db_loader) ------------------------------
    def _apply_patches(self, ty: RustStruct) -> RustType:
        ty2 = self._to_slice(ty)
        if isinstance(ty2, RustStruct):
            ty2 = self._unwrap_argument_type(ty2)
        return ty2

    def _to_slice(self, ty: RustStruct) -> RustType:
        names = {f.name for f in ty.fields}
        if names == {"data_ptr", "length"}:
            data_ptr = next(f.ty for f in ty.fields if f.name == "data_ptr")
            length = next(f.ty for f in ty.fields if f.name == "length")
            if isinstance(data_ptr, RustPointer) and isinstance(length, RustInt) and length.size == self.arch_bytes:
                if ty.name == "&str" and isinstance(data_ptr.pts_to, RustInt) and data_ptr.pts_to.bits == 8:
                    return RustStrRef(size=ty.size)
                return RustSlice(size=ty.size, ele=data_ptr.pts_to)
        return ty

    def _unwrap_argument_type(self, ty: RustStruct) -> RustStruct:
        # core::fmt::rt::Argument { ty: enum { Placeholder{...}, Count(u16) } }
        if ty.name == "core::fmt::rt::Argument" and {f.name for f in ty.fields} == {"ty"}:
            inner = ty.fields[0].ty
            if isinstance(inner, RustEnum):
                for v in inner.variants:
                    if v.name == "Placeholder":
                        return RustStruct(size=ty.size, name="core::fmt::rt::Argument", fields=list(v.fields))
        return ty

    # -- prototypes --------------------------------------------------------
    def _parse_prototype(self, data) -> RustFnProto | None:
        if not data:
            return None
        args = [self._parse_type(a) for a in data.get("args", [])]
        if any(a is None for a in args):
            return None
        ret = self._parse_type(data.get("returnty"))
        return self._fit_abi(args, ret)

    def _fit_abi(self, args: list[RustType], ret: RustType | None) -> RustFnProto:
        """Heuristic Rust ABI fit: aggregates larger than two registers are
        passed/returned by reference (ported from ``_fit_abi``)."""
        big = self.arch_bytes * 2
        new_args: list[RustType] = []
        for a in args:
            if isinstance(a, (RustEnum, RustStruct)) and a.size > big:
                new_args.append(RustPointer(size=self.arch_bytes, pts_to=a))
            else:
                new_args.append(a)
        if isinstance(ret, (RustEnum, RustStruct)) and ret.size > big:
            new_args.insert(0, RustPointer(size=self.arch_bytes, pts_to=ret))
            return RustFnProto(new_args, None, is_retbuf=True)
        return RustFnProto(new_args, ret)
